rust/bridge/shared/macros/src/lib.rs

```diff
@@ -228,15 +242,132 @@ fn jni_name_from_ident(ident: &Ident) -> String {
+        quote!(&mut cx,)
+    } else {
+        quote!()
+    };
+    let input_names = input_args.iter().map(|(name, _ty)| name);
+
+    quote! {
+        #(#input_borrowing)*
+        #(#input_loading)*
+        let __result = #orig_name(#env_arg #(#input_names),*);
+        Ok(node::ResultTypeInfo::convert_into(__result, &mut cx)?.upcast())
+    }
+}
+
+fn node_bridge_fn_async_body(
+    orig_name: &Ident,
+    input_args: &[(&Ident, &Type)],
+    result_kind: ResultKind,
+) -> TokenStream2 {
+    let input_saving = input_args.iter().zip(0..).map(|((name, ty), i)| {
+        let name_arg = format_ident!("{}_arg", name);
+        let name_stored = format_ident!("{}_stored", name);
+        let name_guard = format_ident!("{}_guard", name);
+        quote! {
+            // First, load each argument and save it in a context-independent form.
+            let #name_arg = cx.borrow_mut().argument::<<#ty as node::AsyncArgTypeInfo>::ArgType>(#i)?;
+            let #name_stored = <#ty as node::AsyncArgTypeInfo>::save(&mut cx.borrow_mut(), #name_arg)?;
+            // Make sure we Finalize any arguments we've loaded if there's an error.
+            let mut #name_guard = scopeguard::guard(#name_stored, |#name_stored| {
+                neon::prelude::Finalize::finalize(#name_stored, &mut *cx.borrow_mut())
+            });
+        }
+    });
+
+    let input_unwrapping = input_args.iter().map(|(name, _ty)| {
+        let name_stored = format_ident!("{}_stored", name);
+        let name_guard = format_ident!("{}_guard", name);
+        quote! {
+            // Okay, we've loaded all the arguments; we can't fail from here on out.
+            let mut #name_stored = scopeguard::ScopeGuard::into_inner(#name_guard);
+        }
+    });
+
+    let input_loading = input_args.iter().map(|(name, ty)| {
+        let name_stored = format_ident!("{}_stored", name);
+        quote! {
+            // Inside the future, we load the expected types from the stored values.
+            let #name = <#ty as node::AsyncArgTypeInfo>::load_from(&mut #name_stored);
+        }
+    });
+
+    let env_arg = if result_kind.has_env() {
+        quote!(node::AsyncEnv,)
+    } else {
+        quote!()
+    };
+    let input_names = input_args.iter().map(|(name, _ty)| name);
+
+    let input_finalization = input_args.iter().map(|(name, _ty)| {
+        let name_stored = format_ident!("{}_stored", name);
+        quote! {
+            // Clean up all the stored values at the end.
+            neon::prelude::Finalize::finalize(#name_stored, cx);
+        }
+    });
+
+    quote! {
+        // Use a RefCell so that the early-exit cleanup functions can reference the context
+        // without taking ownership.
+        let cx = std::cell::RefCell::new(cx);
+        #(#input_saving)*
+        #(#input_unwrapping)*
+        Ok(signal_neon_futures::promise(
+            &mut cx.into_inner(),
+            std::panic::AssertUnwindSafe(async move {
```

Comment:
I'm not happy about AssertUnwindSafe, even though we're kind of implicitly relying on it just by working with Neon. I want to come back and see if I can do something safer, but since the stores are all passed `&mut` it might be impossible.

For what it's worth, I put this restriction in the signature of `signal_neon_futures::promise` myself, and I suppose I'm the one taking it back out by putting AssertUnwindSafe into the macro here. :-/